pkg/spec/spec_test.go

```diff
@@ -55,7 +55,7 @@ func TestConflictingFields(t *testing.T) {
 
 			conflictingTags, err := findConflictingJSONTags(reflect.ValueOf(inputStruct))
 			if err != nil {
-				t.Error(errors.Wrap(err, "Unable to find conflicting tags for spec.App"))
+				t.Error(errors.Wrap(err, "Unable to find conflicting tags for spec.DeploymentSpecMod"))
```

Comment:
We can change a code little bit here to make it testable for upcoming controllers, it won't be only for deployment always right?

See I tried following changes:

```diff
diff --git a/pkg/spec/spec_test.go b/pkg/spec/spec_test.go
index 277484e..7262861 100644
--- a/pkg/spec/spec_test.go
+++ b/pkg/spec/spec_test.go
@@ -42,30 +42,45 @@ func TestConflictingFields(t *testing.T) {
 		&Container{},
 		&ConfigMapMod{},
 		&PodSpecMod{},
-		&DeploymentSpecMod{},
 	}
 
-	for _, inputStruct := range structsToTest {
-		t.Run("Testing conflicting fields", func(t *testing.T) {
+	// everytime you add a new controller add a new entry here
+	controllers := []struct {
+		name  string
+		input interface{}
+	}{
+		{
+			name:  "DeploymentSpecMod",
+			input: &DeploymentSpecMod{},
+		},
+	}
 
-			// Checking if input is pointer to struct
-			if err := checkTypePointerToStruct(inputStruct); err != nil {
-				t.Error(errors.Wrap(err, "Input parameter type mismatch"))
-			}
+	for _, controller := range controllers {
+		augmentedStructs := append(structsToTest, controller.input)
 
-			conflictingTags, err := findConflictingJSONTags(reflect.ValueOf(inputStruct))
-			if err != nil {
-				t.Error(errors.Wrap(err, "Unable to find conflicting tags for spec.DeploymentSpecMod"))
-			}
-			if len(conflictingTags) != 0 {
-				t.Logf("The struct %v has unhandled conflicting JSON tags which exist in other structs.", reflect.Indirect(reflect.ValueOf(inputStruct)).Type().String())
-				for tag, structs := range conflictingTags {
-					t.Logf("The JSON tag '%v' exists in %v", tag, structs)
+		for _, inputStruct := range augmentedStructs {
+			t.Run("Testing conflicting fields", func(t *testing.T) {
+
+				// Checking if input is pointer to struct
+				if err := checkTypePointerToStruct(inputStruct); err != nil {
+					t.Error(errors.Wrap(err, "Input parameter type mismatch"))
 				}
-				t.Fatal("Once you handle the above conflicting JSON tag, mark it as handled by adding a 'conflicting' JSON tag to its definition. e.g.\nContainers     []Container `json:\"containers,conflicting,omitempty\"`")
-			}
-		})
+
+				conflictingTags, err := findConflictingJSONTags(reflect.ValueOf(inputStruct))
+				if err != nil {
+					t.Error(errors.Wrapf(err, "Unable to find conflicting tags for spec.%s", controller.name))
+				}
+				if len(conflictingTags) != 0 {
+					t.Logf("The struct %v has unhandled conflicting JSON tags which exist in other structs.", reflect.Indirect(reflect.ValueOf(inputStruct)).Type().String())
+					for tag, structs := range conflictingTags {
+						t.Logf("The JSON tag '%v' exists in %v", tag, structs)
+					}
+					t.Fatal("Once you handle the above conflicting JSON tag, mark it as handled by adding a 'conflicting' JSON tag to its definition. e.g.\nContainers     []Container `json:\"containers,conflicting,omitempty\"`")
+				}
+			})
+		}
 	}
+
 }
 
 // This function takes a map as an input which stores the JSON tag as key and
```

So added a new list of controllers which I append to the main struct `structsToTest` on iteration of each controller. That way we just add a new entry when we add support of jobs and we get it out of the box.

Reply:
But don't forget to add `JobsSpec` there when working on it!

Reply:
A bit confused here.
So, for testing conflicting fields, we need to pass in all the structs in spec.go, so all controller fields and controllers get passed to it. When we add JobSpecMod, it will be just another struct to test.
What do we gain by refactoring this?

Reply:
IIUC you won't be checking the conflicting fields will be checked between `Deployment` and everything else, `Job` and everything else so basically `controller` and everything else so we will have to iterate on list of controllers and everty thing else ?

Reply:
No, conflicting fields has nothing to do with controllers. Basically, every struct in spec.go has to get tested.
This is not related to Kedge's controller logic.

Reply:
oh ok, ignore it then!